src/record/runtime/record.rs

```diff
@@ -0,0 +1,251 @@
+                },
+                Datatype::Int16 => match desc.is_nullable {
+                    true => columns.push(Column::new(
+                        desc.datatype,
+                        desc.name.to_owned(),
+                        Arc::<Option<i16>>::new(None),
+                        desc.is_nullable,
+                    )),
+                    false => columns.push(Column::new(
+                        desc.datatype,
+                        desc.name.to_owned(),
+                        Arc::new(0_i16),
+                        desc.is_nullable,
+                    )),
+                },
+                Datatype::Int32 => match desc.is_nullable {
+                    true => columns.push(Column::new(
+                        desc.datatype,
+                        desc.name.to_owned(),
+                        Arc::<Option<i32>>::new(None),
+                        desc.is_nullable,
+                    )),
+                    false => columns.push(Column::new(
+                        desc.datatype,
+                        desc.name.to_owned(),
+                        Arc::new(0_i32),
+                        desc.is_nullable,
+                    )),
+                },
+            }
+        }
+
+        DynRecord::new(columns, primary_index)
+    }
+}
+
+impl Decode for DynRecord {
+    type Error = RecordDecodeError;
+
+    async fn decode<R>(reader: &mut R) -> Result<Self, Self::Error>
+    where
+        R: Read + Unpin,
+    {
+        let len = u32::decode(reader).await? as usize;
+        let primary_index = u32::decode(reader).await? as usize;
+        let mut columns = vec![];
+        // keep invariant for record: nullable --> Some(v); non-nullable --> v
+        for i in 0..len {
+            let mut col = Column::decode(reader).await?;
+            if i != primary_index && !col.is_nullable {
+                match col.datatype {
+                    Datatype::Int8 => {
+                        let value = col.value.as_ref().downcast_ref::<Option<i8>>().unwrap();
+                        col.value = Arc::new(value.unwrap());
+                    }
+                    Datatype::Int16 => {
+                        let value = col.value.as_ref().downcast_ref::<Option<i16>>().unwrap();
+                        col.value = Arc::new(value.unwrap());
+                    }
+                    Datatype::Int32 => {
+                        let value = col.value.as_ref().downcast_ref::<Option<i32>>().unwrap();
+                        col.value = Arc::new(value.unwrap());
+                    }
+                }
+            }
+            columns.push(col);
+        }
+
+        Ok(DynRecord {
+            columns,
+            primary_index,
+        })
+    }
+}
+
+impl Record for DynRecord {
+    type Columns = DynRecordImmutableArrays;
+
+    type Key = Column;
+
+    type Ref<'r> = DynRecordRef<'r>;
+
+    fn primary_key_index() -> usize {
+        unreachable!("This method is not used.")
```

Comment:
can this method be used to implement another trait such as `StaticRecord` instead of `unreachable!` ?